updates to containerd executor, including basic CNI based network support

ijc opened the pull request:
Apart from actually adding basic CNI support to the containerd executor the largest non-vendoring change here is to refactor some of the network operations into an interface (initially proposed in https://github.com/docker/swarmkit/issues/2230#issuecomment-307402183) in order to provide both CNM and CNI integration.

One nice IMHO result of that refactoring is that all imports of `libnetwork` are now in `manager/allocator/cnmallocator` and `manager/network/cnm` (previously some were also in `controlapi`, which seemed a bit like a layering violation) while all uses of CNI are in `manager/network/cni` and `agent/exec/containerd/adapter.go`. I'd still like to abstract the `adaptor.go` uses further to remove those.

I think there are likely still various bits of `manager/allocator/cnmallocator` which might more properly belong in `manager/network/cnm` (I'm thinking the use of the driver registry for example) but I haven't pulled on that thread more than the amount minimally required here.

The revendoring in `Update containerd to xxxxx` updates a few things which are used outside of the containerd plugin, specifically `github.com/opencontainers/runc` and 
`golang.org/x/sys`. Both look OK to me but I'm not sure if there is some interaction with e.g. vendoring into moby/moby to worry about (especially runc).

The result of this that basic IP level connectivity is made to containers, I've not sorted out thinkgs like service discovery (e.g. DNS) yet.

In terms of testing I've been using the [swarmd project in LinuxKit](https://github.com/linuxkit/linuxkit/tree/master/projects/swarmd) to fire up a cluster of VMs running [Weave Net](https://www.weave.works/oss/net/) based CNI networking. A network is created with `swarmctl network create --cni '{ "name": "weave", "type": "weave-net" }'` and then `service create ... --network weave` can be used in the normal way.

aaronlehmann commented on manager/network/model.go:
```diff
@@ -0,0 +1,15 @@
+	Allocator() (networkallocator.NetworkAllocator, error)
+	SetDefaults(spec *api.NetworkSpec) error
+	ValidateNetworkSpec(spec *api.NetworkSpec) error
+	PredefinedNetworks() []networkallocator.PredefinedNetworkData
```
I feel like the definition of `PredefinedNetworkData` belongs in this package rather than `networkallocator`.

We can't move it on its own because that would cause a circular dependency.

I wonder if it would also make sense to move the `NetworkAllocator` interface definition here. Then this package would contain the interface definitions, and `networkallocator` would have the implementations.

I'm not completely sure either way, just throwing this out as an idea. Something doesn't feel quite right about this package depending on the allocator.

ijc replied:
That's a good point/idea. I'll try it.

ijc replied:
Actually `PredefinedNetworkData` isn't used anywhere in the network allocator package after the changes in this PR so moving it makes absolute sense and doesn't cause a circular dependency.

I'm still considering possibly moving `networkallocator.NetworkAllocator` to `network.Allocator`, I think I like it though.

ijc replied:
@aaronlehmann done. I also refactored the rest of `manager/allocator/networkallocator` and a few bits of `manager/allocator/network.go` out into various more suitable places, so the packages which depend on the allocator are now (IMHO at least) much cleaner.

aaronlehmann commented on manager/network/model.go:
```diff
@@ -0,0 +1,15 @@
+	SetDefaults(spec *api.NetworkSpec) error
+	ValidateNetworkSpec(spec *api.NetworkSpec) error
+	PredefinedNetworks() []networkallocator.PredefinedNetworkData
+	SupportIngressNetwork() bool
```
`SupportsIngressNetwork`?

aaronlehmann replied:
Another way to do this is only define an `IngressNetwork` method on implementations that support the ingress network, and type-assert to check whether a particular implementation has this method or not. This keeps implementation-specific details from creeping into the main interface definition. It can be more cumbersome to use, though.

ijc replied:
I'll try this too.

ijc replied:
@aaronlehmann I made the name change as an interim measure. but WRT the `IngressNetwork` method I think it might make sense to move the `newIngressNetwork` function into the cnm version of that and use the type check WDYT?

I'm also eyeing up `newPredefinedNetwork` in a similar light, in essence making `PredefinedNetworks` return an `api.Network` instead of the `PredefinedNetworkData` and so embedding the function into CNM where it is used.

aaronlehmann commented on manager/network/model.go:
```diff
@@ -0,0 +1,15 @@
+
+// Model is an abstraction over the Network Model to be used.
+type Model interface {
+	Allocator() (networkallocator.NetworkAllocator, error)
```
I would call this `NewAllocator`.

aaronlehmann commented on agent/exec/containerd/controller.go:
```diff
@@ -57,6 +57,32 @@ func (r *controller) ContainerStatus(ctx context.Context) (*api.ContainerStatus,
+			continue
+		}
+		for _, ip := range ns.cni.IPs {
+			log.G(ctx).Debugf("Iface# %d, num interfaces %d\n", ip.Interface, len(ns.cni.Interfaces))
```
no `\n`.

aaronlehmann commented on manager/allocator/allocator_test.go:
```diff
@@ -716,6 +718,139 @@ func isValidSubnet(t assert.TestingT, subnet string) bool {
+	netWatch, cancel := state.Watch(s.WatchQueue(), api.EventUpdateNetwork{}, api.EventDeleteNetwork{})
+	defer cancel()
+	taskWatch, cancel := state.Watch(s.WatchQueue(), api.EventUpdateTask{}, api.EventDeleteTask{})
+	defer cancel()
```
I believe this works as intended, but it might clarify the code to avoid reusing the name `cancel`.

ijc commented:
Addressed all comments except https://github.com/docker/swarmkit/pull/2299#pullrequestreview-48433570 which GH has helpfully decided was old due to my touching the line for other reasons. Leaving this comment as a reminder to self.

aaronlehmann commented on api/network.go:
```diff
@@ -0,0 +1,11 @@
+package api
+
+// IsIngressNetwork check if the network is an ingress network
+func IsIngressNetwork(nw *Network) bool {
```
Could you put these helper functions in an `api/network` package? We've tried to reserve `api` for autogenerated code, though there are some exceptions now.

aaronlehmann replied:
I also think ``github.com/docker/swarmkit/manager/network` would be a fine place to put these. In particular, `IsIngressNetworkNeeded` is basically business logic, so the `manager/network` package might make more sense.

aaronlehmann commented:
Mostly LGTM. Left a few comments.

aaronlehmann commented on api/types.proto:
```diff
@@ -459,12 +459,26 @@ enum TaskState {
+	string interface = 2;
+
+	// Address is the IP address in the form ipv4:XXX or ipv6:YYY
+	string address = 3;
```
Wondering if it would make sense to break apart the protocol and address into separate fields.

If this format is borrowed from CNI and is well-defined there, this is probably fine, but in general I like to avoid microformats.

ijc replied:
Good point, I did indeed make this up myself.

`api.EndPoint.VirtuallIP.addr` is a string. Looking at the usage it appears to be defined as one which is compatible with `net.ParseCIDR`/`net.ParseIP`. This seems like a reasonable enough precedent to follow.

ijc replied:
Related to this I don't really like what I've done wrt error reporting at https://github.com/ijc/swarmkit/blob/9c619b516e62f103df460dfc8ffd9652634733ac/agent/exec/containerd/controller.go#L66, essentially it just uses an "error:" prefix to distinguish.

I think maybe it would be better to have a separate error field, probably as a `oneof`.

aaronlehmann commented on manager/network/cnm/cnm.go:
```diff
@@ -0,0 +1,149 @@
+	pg plugingetter.PluginGetter
+}
+
+// New produces a fresh CNI Network Model
```
CNM

aaronlehmann commented on manager/network/cnm/cnm.go:
```diff
@@ -0,0 +1,149 @@
+		if cnmallocator.IsBuiltInDriver(driver.Name) {
+			return nil
+		}
+	default:
```
Remove the empty `default` case.

ijc commented:
I've now addressed the outstanding review (sorry for the delay), I intend to fold those changes back into the right place in the course of the rebase which I need to do also.

I'm also going to pull in an update containerd v1.0.0-alpha1 (or later if there's another release in the meantime) at the same time.

ijc commented:
Rebased. Only minor hiccup was the need to plumb a `PluginGetter` back through `node.Config` and `manager.Config` (previously "Abstract Network Model into an interface" managed to remove this) since the dispatcher now uses it too (since e9a7bc02444ede1505b423440e7ef3b2468334a1).

aaronlehmann commented on agent/exec/containerd/adapter.go:
```diff
@@ -193,6 +214,29 @@ func (c *containerAdapter) isPrepared() bool {
 }
 
+func cniConfig(n *api.Network) (*libcni.NetworkConfig, error) {
+	if n.DriverState.Name != "cni" {
```
Could `DriverState` be nil here?

ijc replied:
`cni.ValidateNetworkSpec` would have rejected it if it were, and `cniNetworkAllocator.Allocate` ensures (redundantly) that it is not, but I think I shouldn't rely on the behaviour of such distant components so I will add a check here too.

ijc replied:
Oops, `cni.ValidateNetworkSpec` is looking at `DriverConfig` not `DriverState`, so `cniNetworkAllocator.Allocate` isn't redundant, I should still add the check here though.

aaronlehmann commented on manager/network/cni/cni.go:
```diff
@@ -0,0 +1,84 @@
+	}
+
+	if spec.Annotations.Name == "" {
+		spec.Annotations.Name = cni.Network.Name
```
Just noticed this bit. It's sort of problematic - controlapi shouldn't modify the spec, which should only be user data. I can't think of any horrible implications from this particular statement, but we've worked pretty hard to keep a clear separation between user-controlled data and fields that are modifiable by swarmkit, and I wouldn't want to introduce an exception without a very good reason. Among other things, we're thinking of adding a mechanism for signing specs before they're submitted to swarmkit, in the future.

Could we remove `SetDefaults` and instead dynamically resolve an empty`spec.Annotations.Name` field?

ijc replied:
This could have far reaching implications, for example places like `validateAnnotations` would need a special case which would depend on the type of spec the annotation was embedded in and we would need to arrange that uses of the network spec always got their name via a wrapper function instead of going to it directly.

There's also the overhead of having to parse the JSON each time (although I have an open plan to try caching that).

I was hoping to avoid requiring the user to specify both `spec.Annotations.Name` and the cni name identically, but perhaps I should flip this around and instead allow the cni name to be empty and fill it in from the (required) annotation name. That would "just" require me to figure out how to round trip an arbitrary JSON object through a parser to add a field without knowing in advance all of the fields which are present. I'm not actually sure if the go json library can do that, but I'll investigate.

Or another alternative would be to add a `Name` field to the `api.Network` object which contains the canonical name as determined by looking at the spec and the cni. AIUI that object is, unlike the spec, internal and fair game for mutating.

WDYT?

aaronlehmann replied:
> I was hoping to avoid requiring the user to specify both spec.Annotations.Name and the cni name identically

I think it's not so bad to require the user to specify `spec.Annotations.Name`. I think we require non-empty values for some object types.

I'm not sure I understand why `spec.Annotations.Name` would have to match the CNI name, but if they do, enforcing this constraint in controlapi validation sounds okay to me.

ijc replied:
> I'm not sure I understand why spec.Annotations.Name would have to match the CNI name

Just to avoid confusion since the CNI name might get used where the admin might see it e.g. the bridge name (I think) and having that match swarmd's idea of the network's name seems useful.

Maybe it's not worth enforcing it though, if an admin wants to confuse themself that's up to them.

aaronlehmann replied:
I'd expect client software (i.e. Docker) to use the same value in each field in practice. I don't think it's necessary to enforce it at the swarmkit level.

ijc replied:
Okey dokey, it certainly simplifies my life here ;-)

aaronlehmann commented on api/types.proto:
```diff
@@ -465,12 +465,31 @@ enum TaskState {
 }
 
+// State of Networks attached to a container
+message ContainerNetworkStatus {
```
I noticed that this would be used in flat list, where the same values of `network` and `interface` may be repeated to provide multiple addresses.

If one of the networks is showing an error, there would be a single entry for it with an error instead of an IP.

This seems different from the way networks are represented internally in the executor. There's a set of networks at the top level, and each network can have zero or more IPs underneath it. Would it make sense to replicate that hierarchy here? My first thought is it would be a little less confusing than the flattened list here with the `oneof` choosing between an IP or an error condition, but I'm not really familiar with CNI, so it's very possible I'm missing something.

ijc replied:
I have literally no idea why I chose to flatten this in the way I did, I probably started down one path before I had fully groked the CNI data structures and never reevaluated the base decision.

I'll do as you suggest and unflatten it making it a (non-`oneof`) `string error`  and `repeated string ip`.

ijc replied:
So, this is a bit more complicated than I had remembered. The result of a single CNI add call (corresponding to a single `api.Tasks.Network`) is a [Result struct](https://github.com/containernetworking/cni/blob/master/pkg/types/current/types.go#L135..L142) which contains (among other things) an array of [`Interfaces`](https://github.com/containernetworking/cni/blob/master/pkg/types/current/types.go#L240..L244) (name, sandbox, max) and an array of [IPs](https://github.com/containernetworking/cni/blob/master/pkg/types/current/types.go#L257..L264) (4 vs 6, reference to the Interfaces array, network and gateway). There's also routes and dns in there. There's a (brief) example under the "Result" heading in [the spec](https://github.com/containernetworking/cni/blob/master/SPEC.md).

There isn't a requirement that `len(Interfaces) == 1`, i.e. a single CNI add call might produce multiple interfaces. It's also possible that `len(Interfaces) < len(IPs)`, i.e. that a single interface might be given multiple IP addresses. There's also more information here (gateways, routes etc) than I have exposed so far.

I'm not really sure how best to approach this. I'm leaning towards a more hierarchical representation of the raw CNI (in particular dereferencing the interface array index) and leaving out the other fields I mentioned for now (they can be added later easily enough). So something of the form:
```
message ContainerNetworkStatus {
	// Network is the Network ID
	string network = 1;

	// Error, if any.
	string error = 2;

	message NetworkInterface {
		string name = ??;

		// Each IP is in dotted decimal ("192.0.2.1") or IPv6 ("2001:db8::68") form.
		repeated string ip = ?? [(gogoproto.customname) = "IPs"];
	};

	repeated NetworkInterface interfaces = ??;
}
```

While I was here I did spot that my construction of an interface name (`eth%d`) to pass to the CNI add call was unnecessary (and probably wrong). The spec requires the plugin to use that name if it is given, but if not can choose its own name(s), which seems like the correct approach to use (and is reflected in the state above).

aaronlehmann replied:
This proposal makes sense to me. Again, I'm not really familiar with CNI, but this is probably what I'd come up with if asked to abstractly describe network interfaces and their IPs.

ijc commented:
Review has been addressed.

aaronlehmann commented on api/types.proto:
```diff
@@ -470,16 +470,17 @@ message ContainerNetworkStatus {
-		string error = 4;
-	}
+		// Each IP is in dotted decimal ("192.0.2.1") or IPv6 ("2001:db8::68") form.
+		repeated string ip = 2 [(gogoproto.customname) = "IPs"];
```
The custom name isn't consistent with the protobuf name - only one is plural.

I think I'd just go with `IP`.

aaronlehmann commented on agent/exec/containerd/controller.go:
```diff
@@ -57,34 +57,56 @@ func (r *controller) ContainerStatus(ctx context.Context) (*api.ContainerStatus,
+						ifnum = 0
+					} else {
+						// This cannot happen given the workaround above, but keep this check for future use.
+						l.Debugf("Interface index %d out of range, but no interfaces", ip.Interface)
```
"but no interfaces"?

ijc replied:
I'm trying to express that we cannot fallback to `ifnum = 0` because there are zero interfaces. How about `"Interface index %d out of range, no fallback available"`?

aaronlehmann commented on manager/network/cni/cni.go:
```diff
@@ -66,8 +51,7 @@ func (nm *cni) ValidateNetworkSpec(spec *api.NetworkSpec) error {
 	if spec.Annotations.Name != cni.Network.Name {
-		return grpc.Errorf(codes.InvalidArgument,
-			"CNI Network name (%q) must match Spec annotations name (%q)",
+		log.G(ctx).Infof("CNI Network name (%q) does not match Spec name (%q)",
```
Seems okay to log the mismatch, though I'm not sure users will really notice this in the manager log.

ijc replied:
You are most likely correct but it seemed better than nothing.

aaronlehmann commented:
There's a unit test compilation issue :(.

ijc commented:
Review (and test failure) addressed.

aaronlehmann commented:
LGTM

ijc commented:
CI fail is:
```
command cd "$WORKDIR" && TMPDIR=/tmpfs make ci took more than 10 minutes since last output
```
I don't think this is specific to the changes here.

ijc commented:
Force pushed to resolve conflict with 82c36fe16ec5a4400fa8b4c36b883d509500360a.
I'm going to tag on some commits to update to latest containerd alpha too.

ijc commented:
Ah, I remember why I hadn't done alpha2 (or now alpha3), it's the logrus change.

I'll revisit that aspect once #2334 has landed. I would expect that merging this PR would make #2334 a bit easier since the delta from alpha1 (here) to alpha2,3,etc should be small since this code uses the client library.

I'll do a fake/local rebase of this onto #2334 to check that theory.

/cc @dmcgowan.

AkihiroSuda commented:
@ijc What's current status and plan?

ijc commented:
I'm afraid I've moved onto other things and have no current plans to pursue this further.
